discord/voice_client.py

```diff
@@ -56,9 +58,89 @@
 from . import opus
 from .backoff import ExponentialBackoff
 from .gateway import *
-from .errors import ClientException, ConnectionClosed
+from .errors import ClientException, ConnectionClosed,  InvalidVoicePacket
 from .player import AudioPlayer, AudioSource
 
+class VoicePacket:
+    """Represents a packet of voice data.
+
+    You should not construct these yourself, they will be constructed from
+    incoming UDP packets on the socket used for voice data.
+
+    Attributes
+    ----------
+    sequence : int
+        The sequence id for the packet. Must be between 0 and 65535. Each
+        packet has a value of 1 greater than the previous packet from the
+        same source. After the sequence reaches the maximum value (65535),
+        the next packet has a sequence value of 0.
+    timestamp : int
+        The timestamp for the packet.
+    ssrc : int
+        The id of the user (or source) of this voice packet.
+    buff : bytes
+        The Opus packet (as bytes).
+    user : Optional[:class:`abc.User`]
+        The user who the packet belongs to
+    pcm : Optional[bytes]
+        The decoded PCM data.
+    """
+    # Struct format of the packet metadata
+    _FORMAT = '>HHII'
+    # Packets should start with b'\x80\x78' (32888 as a big endian ushort)
+    _CHECK = 32888
+    # Some packets will start with b'\x90\x78' (36984)
+    _CHECK2 = 36984
+
+    def __init__(self, sequence=None, timestamp=None, ssrc=None, buff=None, user=None):
+        self.sequence = sequence
+        self.timestamp = timestamp
+        self.ssrc = ssrc
+        self.buff = buff
+        self.user = user
+
+    @classmethod
+    def unpack(cls, packet, voice_client):
+        if len(packet) < 13:
+            raise InvalidVoicePacket('packet is too small: {}'.format(packet))
+
+        # Unpack header
+        check, seq, ts, ssrc = struct.unpack_from(cls._FORMAT, packet)
+        header = packet[:12]
+        buff = packet[12:]
+
+        # Check the packet is valid
+        if check != cls._CHECK and check != cls._CHECK2:
+            fmt = 'packet has invalid check bytes: {}'
+            raise InvalidVoicePacket(fmt.format(packet))
+
+        # Decrypt data
+        nonce = bytearray(24)
+        nonce[:12] = header
+        box = nacl.secret.SecretBox(bytes(voice_client.secret_key))
+        buff = box.decrypt(bytes(buff), bytes(nonce))
+
+        # Packets starting with b'\x90' need the first 8 bytes ignored
+        if check == cls._CHECK2:
+            buff = buff[8:]
+
+        # Lookup the SSRC and then get the user
+        user_id = 0
+        if voice_client.channel.id in voice_client._ssrc_lookup:
+            if ssrc in voice_client._ssrc_lookup[voice_client.channel.id]:
```

Comment:
it may be worth holding onto packets that you can't get a valid user ID for until the speaking event arrives from the websocket.